required/latex-lab/latex-lab-toc-kernel-changes.dtx

```diff
@@ -0,0 +1,429 @@
+% |hyperref| for the bookmark levels. While headings are more a class thing, it
+% could make sense to provide a standard set already in the kernel.
+% The numbers are a bit different to the one in hyperref, but imho sensible.
+%    \begin{macrocode}
+\def\toclevel@part{-1}
+\def\toclevel@chapter{0}
+\def\toclevel@section{1}
+\def\toclevel@subsection{2}
+\def\toclevel@subsubsection{3}
+\def\toclevel@paragraph{4}
+\def\toclevel@subparagraph{5}
+\def\toclevel@figure{1}
+\def\toclevel@table{1}
+%    \end{macrocode}
+%
+% \subsection{\cs{contentsline} extension}
+% [latex-lab]
+% The tagging code must add code at the begin and end of \cs{contentsline}.
+% This code must have access to the arguments. We set it up as config point
+% for now.
+%    \begin{macrocode}
+\def\@contentsline@cfgpoint@before#1#2#3#4{}
+\def\@contentsline@cfgpoint@after#1#2#3#4{}
+\def\@contentsline@arg@one{}%level
+\def\@contentsline@arg@two{}%text
+\def\@contentsline@arg@three{}%page
+\def\@contentsline@arg@four{}%destination  
+\def\contentsline#1#2#3#4%
+  {%
+    \def\@contentsline@arg@one{#1}%level
+    \def\@contentsline@arg@two{#2}%text
+    \def\@contentsline@arg@three{#3}%page
+    \def\@contentsline@arg@four{#4}%destination  
+    \@contentsline@cfgpoint@before{#1}{#2}{#3}{#4}%
+    %\gdef\@contentsline@destination{#4}%
+    \csname l@#1\endcsname{#2}{#3}%
+    \@contentsline@cfgpoint@after{#1}{#2}{#3}{#4}%
+  }
+%    \end{macrocode}
+%
+% \subsection{\cs{addcontentsline} changes}
+% [latex-lab]
+%
+% \cs{addcontentsline} should always write the unique counter representation
+% \cs{@currentHref} so that we can create the |/Ref| key. At the same time
+% we need a hook, so that hyperref can add its bookmarks code. The generic hook
+% would be ok, but we need the arguments, so we store them first.
+%
+%    \begin{macrocode}
+\NewHook{cmd/addcontentsline/before}
+\def\addcontentsline#1#2#3{% toc extension, type, tag
+    \def\addcontentsline@arg@one{#1}% name??
+    \def\addcontentsline@arg@two{#2}%
+    \def\addcontentsline@arg@three{#3}%
+    \UseHook{cmd/addcontentsline/before}%
+    \addtocontents{#1}{%
+      \protect\contentsline{#2}{#3}{\thepage}{\@currentHref}\protected@file@percent
+    }}
+%    \end{macrocode}
+%
+% \subsection{\cs{@starttoc}}
+% [kernel?]
+%
+% We add a configuration point before and after reading the file
+% for the tagging.
+%    \begin{macrocode}
+\def\@starttoc@cfgpoint@before#1{}
+\def\@starttoc@cfgpoint@after#1{}
+\def\@starttoc#1{%
+  \begingroup
+    \makeatletter
+    \@starttoc@cfgpoint@before{#1}%
+    \@input{\jobname.#1}%
+    \@starttoc@cfgpoint@after{#1}
+    \if@filesw
+      \expandafter\newwrite\csname tf@#1\endcsname
+      \immediate\openout \csname tf@#1\endcsname \jobname.#1\relax
+    \fi
+    \@nobreakfalse
+  \endgroup}
+%    \end{macrocode}
+%
+% \subsection{Formatting of the toc entries}
+%
+% \subsubsection{Kernel commands}
+% [latex-lab] or [kernel?]?
+%
```

Comment:
all of this should become a set of template (next step)

Reply:
right, have overlooked that. So maybe that is what should be put into some further kernel code update file instead of doing it manually here.